File: 2020-algorithms-studying/Arrays/sortColors.js
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var sortColors = function(nums) {
    //IMplement a sorting algorithm from scratch
    const swap = (i,j) => {
        const temp = nums[j]
        nums[j] = nums[i]
        nums[i] = temp
        
    }
    
    for(let i = 0;i<nums.length;i++){
        for(let j = 0;j<nums.length;j++){
            if(nums[i] < nums[j]){
                swap(i,j)
            }
        }
    }
};

File: sort-colors.py
class Solution:
    def sortColors(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        
        
        Approaches
        
        1. Bucket sort
        
            [0,0,0]
             0,1,2
             
        Iterate through array
        Count each color
        Iterate through array and replace colors
        
        
        2. Three Pointers
        
        
        [2,0,2, 1,1, 0]
                     blue
         white       
                      
        """
        
        
        buckets = [0,0,0]
        
        for val in nums:
            buckets[val] += 1
        
        
        pt = 0
        
        for i in range(len(nums)):
            
            while buckets[pt] == 0:
                pt += 1
            nums[i] = pt
            buckets[pt] -= 1
            
            
            
        
        
        